### PR TITLE
[1/N] [Dispatchable Collectives] Create Backend class

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations
* #83876 [6/N] [Dispatchable Collectives] Update recv with CPU / CUDA implementations
* #83859 [5/N] [Dispatchable Collectives] Update send with CPU / CUDA implementations
* #83810 [4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations
* #83735 [3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations
* #83680 [2/N] [Dispatchable Collectives] Extract ProcessGroup::Work into a separate class and update references
* **#83679 [1/N] [Dispatchable Collectives] Create Backend class**

### Changes:
- Create a new Backend class which contains collectives similar to that of https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroup.hpp.

### Context

https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D38839213](https://our.internmc.facebook.com/intern/diff/D38839213)

### CODE REVIEW

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38839213).

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38839213).

**Comment by vadimkantorov** [comment]
This would be nice because even in GPU training sometimes broadcasting CPU tensor is needed. And in reverse, for CPU-only inference, a different backend should be used...

File: torch/csrc/distributed/c10d/Backend.hpp [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+  // for GLOO and NCCL backends currently.
+  virtual void setSequenceNumberForGroup() {
+    auto backendName = getBackendName();
+    TORCH_CHECK(
+        false,
+        c10::str(
+            "Backend ",
+            backendName,
+            " does not yet support sequence numbers."));
+  }
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/Backend.hpp:210]: Just for our record: we'd need to think whether `SequenceNumber` related methods should stay here or be moved to `ProcessGroup`. For the purpose of smooth refactorization, they can stay here for now.

**Review by kwen2501 — APPROVED** [review]
LGTM! Thanks!

File: torch/csrc/distributed/c10d/Backend.hpp (hunk 1 above)
Comment (H-Huang) [reply torch/csrc/distributed/c10d/Backend.hpp:210]: Makes sense, agree on this!

**Pull request closed without merging**